site/_posts/2020-07-23-securing-react-native-with-oauth.md

```diff
@@ -0,0 +1,480 @@
+```
+
+Launch the ngrok proxy. For full details, consult [the documentation](https://ngrok.com/docs). For our purposes, running it from the command line is sufficient. We want to start HTTP tunnel forwarding to `localhost` and port `9011`, where FusionAuth is listening. To do so, run the following command:
+
+```shell
+./ngrok http 9011
+```
+
+We'll get a random URL which forwards traffic to our FusionAuth instance. It'll be something like `https://ce2f267ff5a5.ngrok.io`. We can reference this value from our Android device and any traffic sent will be forwarded. We will also use this for our iOS app for consistency, even though the iOS emulator can connect to localhost.
+
+Now, we can move on to coding.
+
+## The implementation
+
+As always, if you want to skip ahead, grab the code from the [GitHub repository](https://github.com/fusionauth/fusionauth-example-react-native).
+
+First, we need to modify the `App.js` file. Add necessary imports as shown in the code snippet below:
+
+```javascript
+//...
+import React, { useState, useCallback, useMemo } from 'react';
+import { Alert } from 'react-native';
+import { authorize, refresh, revoke, prefetchConfiguration } from 'react-native-app-auth';
+//...
+```
+
+Next, we need to create a `configs` object as shown in the code snippet below:
+
+```javascript
+//...
+const configs = {
+  fusionauth: {
+    issuer: 'https://ce25267ff5a5.ngrok.io', 
+    clientId: '253eb7aa-687a-4bf3-b12b-26baa40eecbf',
+    redirectUrl: 'fusionauth.demo:/oauthredirect',
+    additionalParameters: {},
+    scopes: ['offline_access'],
+  }
+}
+//...
+```
+
+Here's more information about the properties of the `fusionauth` object in the `configs` object:
+
+* `issuer` is the URL for the FusionAuth server.
+* `clientId` is the ID that we grabbed from the FusionAuth dashboard.
+* `redirectURL` is the URL that we set up before, with a callback path defined by the react native app auth library.
+
+Make sure you update the `issuer` and `clientId` keys in this object with your configuration values.
+
+We also add any additional parameters we need to pass (none, in this case), as well as scopes to request. We're requesting the `offline_access` scope as that will return a `refresh_token`, which can be used to request additional access tokens.
+
+Then, create a default auth state in order to handle the response from the server, as shown in the code snippet below:
+
+```javascript
+//...
+const defaultAuthState = {
+  hasLoggedInOnce: false,
+  provider: '',
+  accessToken: '',
+  accessTokenExpirationDate: '',
+  refreshToken: ''
+};
+
+const [authState, setAuthState] = useState(defaultAuthState);
+//...
+```
+
+Now, we are ready to configure authorization.
+
+## Configuring Authorization
+
+We need to create a function to perform authorization. This will require the `configs` object previously created. The implementation of the function is below:
+
+```javascript
+//...
+const handleAuthorize = useCallback(
```

Comment:
I'm not sure where this code goes or how it hooks up to anything. The explanation of how this all works is pretty light. Maybe all the packages that are installed do the magic, but I'm lost at this point.

Reply:
I pulled in the entire `App.js` file and added more context around this.